opwen_email_server/actions.py

```diff
@@ -238,6 +239,38 @@ def _new_email_id(cls, email: str) -> str:
         return sha256(email.encode('utf-8')).hexdigest()
 
 
+class ProcessServiceEmail(_Action):
+    def __init__(self,
+                 raw_email_storage: AzureTextStorage,
+                 email_storage: AzureObjectStorage,
+                 next_task: Callable[[str], None],
+                 email_parser: Callable[[str], dict] = None,
+                 email_formatter: Callable[[str], dict] = None):
+
+        self._raw_email_storage = raw_email_storage
+        self._next_task = next_task
+        self._email_parser = email_parser or MimeEmailParser()
+        self._email_formatter = email_formatter or EmailFormatter()
+
+    def _action(self, resource_id):  # type: ignore
+        try:
+            mime_email = self._raw_email_storage.fetch_text(resource_id)
+        except ObjectDoesNotExistError:
+            self.log_warning('Inbound email %s does not exist', resource_id)
+            return 'skipped', 202
+
+        email = self._email_parser(mime_email)
+        self._raw_email_storage.delete(resource_id)
```

Comment:
One nice property we get via queues is free retries: in case the processing crashes somewhere before the end of the method, the message that triggered the workload isn't deleted from the queue and we'll retry the processing with the next worker.

To make the retries work, we need to delay any stateful actions (such as deleting the email) until the end of the method to make sure that if the action gets retried, all required resources are still there.

I assume you copied this approach from somewhere in the code? If so: that place is also doing it wrong and should be updated.

Reply:
Oh, interesting! I just remembered seeing the delete method being used but I didn't really think much about the placement of it. I see now that delete is mostly being used at the end. The one place where delete is being used before calling next task is in [StoreInboundEmails](https://github.com/ascoderu/opwen-cloudserver/blob/ef72c504c29e3776741e587b4e62041dc69caa90/opwen_email_server/actions.py#L70)